eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/ErrorHandlers.scala

```diff
@@ -268,13 +274,31 @@ trait ErrorHandlers extends CommonHandlers {
-      case negotiating: DATA_NEGOTIATING => DATA_CLOSING(d.commitments, fundingTx = None, waitingSince = nodeParams.currentBlockHeight, negotiating.closingTxProposed.flatten.map(_.unsignedTx), nextRemoteCommitPublished = Some(remoteCommitPublished))
+      case negotiating: DATA_NEGOTIATING => DATA_CLOSING(d.commitments, fundingTx = None, waitingSince = nodeParams.currentBlockHeight, alternativeCommitments = Nil, mutualCloseProposed = negotiating.closingTxProposed.flatten.map(_.unsignedTx), nextRemoteCommitPublished = Some(remoteCommitPublished))
       // NB: if there is a next commitment, we can't be in DATA_WAIT_FOR_FUNDING_CONFIRMED so we don't have the case where fundingTx is defined
-      case _ => DATA_CLOSING(d.commitments, fundingTx = None, waitingSince = nodeParams.currentBlockHeight, mutualCloseProposed = Nil, nextRemoteCommitPublished = Some(remoteCommitPublished))
+      case _ => DATA_CLOSING(d.commitments, fundingTx = None, waitingSince = nodeParams.currentBlockHeight, alternativeCommitments = Nil, mutualCloseProposed = Nil, nextRemoteCommitPublished = Some(remoteCommitPublished))
     }
     goto(CLOSING) using nextData storing() calling doPublish(remoteCommitPublished, d.commitments)
   }
 
+  def handleRemoteSpentAlternative(commitTx: Transaction, alternativeCommitments: List[Commitments], d: PersistentChannelData) = {
+    val commitments_opt = alternativeCommitments.find(c => c.remoteCommit.txid == commitTx.txid || c.remoteNextCommitInfo.left.toOption.exists(_.nextRemoteCommit.txid == commitTx.txid))
+    require(commitments_opt.nonEmpty, "there should be a commit spending an alternative funding tx matching this transaction")
+    val commitments = commitments_opt.get
+    log.warning("they published their commit with txid={} spending an alternative funding tx with fundingTxid={}", commitTx.txid, commitments.commitInput.outPoint.txid)
+
+    context.system.eventStream.publish(TransactionPublished(d.channelId, remoteNodeId, commitTx, Closing.commitTxFee(commitments.commitInput, commitTx, commitments.localParams.isInitiator), "remote-commit"))
+    // We wait for this alternative funding tx to be confirmed before claiming the commit tx outputs, as it could still
+    // be replaced by another funding tx.
+    blockchain ! WatchFundingConfirmed(self, commitments.commitInput.outPoint.txid, nodeParams.channelConf.minDepthBlocks)
+    val nextData = d match {
+      case waitForFundingConfirmed: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED => DATA_CLOSING(d.commitments, waitForFundingConfirmed.signedFundingTx_opt, waitingSince = nodeParams.currentBlockHeight, alternativeCommitments, mutualCloseProposed = Nil)
+      case closing: DATA_CLOSING => closing
+      case _ => DATA_CLOSING(d.commitments, None, waitingSince = nodeParams.currentBlockHeight, alternativeCommitments, mutualCloseProposed = Nil)
+    }
+    goto(CLOSING) using nextData storing()
+  }
```

Comment:
I don't think this method is necessary: we only need to track the funding tx, until one of them reaches min_depth, which eliminates all alternative commitments. From then, we are back to the nominal case. Our system already supports reacting asynchronously to transactions, this is what we do when the node was offline during a force close.

Reply:
> we only need to track the funding tx

You mean funding txs, right? Note that for (future) splicing, we will have something similar, but in the `NORMAL` state, where we'll have a set of future funding txs that spend the current txs.

> we only need to track the funding tx, until one of them reaches min_depth, which eliminates all alternative commitments.

I don't understand how you would achieve that if you remove this method...in OFFLINE and SYNCING you need to identify cases where `WatchFundingSpentTriggered` is for an alternative commitment, to go to the closing state and put a `WatchConfirmed` on that tx? Remember that while offline, we don't have a watch-confirmed on those transactions, we only set these watches when receiving the remote `channel_reestablish`.

Reply:
We leave the handlers in `OFFLINE`/`SYNCING`, but make them call `spendLocalCurrent` (after all this is an error scenario). Then we rely on the `WAIT_FOR_FUNDING_CONFIRMED` handler in `CLOSING` to deal with the commitment switch?

Reply:
I think you're right, this should work. I implemented this in https://github.com/ACINQ/eclair/pull/2274/commits/7b7bfac5bb44092cac0a7439723abff63ae4f356 let me know if I missed something. There is duplication in the handlers in OFFLINE and SYNCING, but I'm not really sure how to name the method we'd use to factor those together, do you have an idea?

Reply:
So here are the possible transitions (in a RBF scenario):
- `DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED`->`CLOSING`
_channel was created then closed, no disconnections_
  In this case, we have put a `WatchFundingConfirmed` and a `WatchFundingSpent` on all alternate funding txs in state `WAIT_FOR_DUAL_FUNDING_CREATED`. There will be a harmless race in state `CLOSING` between `WatchFundingSpentTriggered` (which cause us to put a duplicate `WatchFundingConfirmed`, deduped by the watcher) and `WatchFundingConfirmedTriggered`. Btw I think we should also handle the nominal case were the latest funding tx confirms, and clean up the alternate commitments.
- `WAIT_FOR_FUNDING_CREATED`->`OFFLINE`[->`SYNCING`]->`CLOSING`
  _channel was created then disconnected, then closed, but eclair stayed up the whole time_
  Same as above, except that  the `WatchFundingConfirmed` and `WatchFundingSpentTriggered` that were put in `WAIT_FOR_DUAL_FUNDING_CREATED` may also race in `OFFLINE`/`SYNCING`. The result is the same, we may put duplicate `WatchFundingConfirmed` (in that case they will be ignored), but it's harmless.
- `DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED`-><FSM stopped>->`WAIT_FOR_INIT_INTERNAL`[->`OFFLINE`][->`SYNCING`]->`CLOSING``
  _channel was created then eclair went offline, then tge channel closed_
  In this scenario, we have not put any `WatchFundingConfirmed`, so there won't be a race: we will (1) receive `WatchFundingSpentTriggered`, then (2) put `WatchFundingConfirmed`, then (3) receive `WatchFundingConfirmedTriggered`. ⚠️ Note that we currently have a bug where we don't put a `WatchFundingSpent` on alternative funding txs in `WAIT_FOR_INIT_INTERNAL`.

I think we could simplify the flow and not rely on `WatchFundingSpent` to "discover" alternate funding txs. This approach was justified when there was a huge list of potential spending txs, and we wanted to limit the number of `WatchConfirmed`, which are costly. But in this case, where there is only a handful of competing funding txs, we could only rely on `WatchConfirmed`. We would then use a catch-all handler in `whenUnhandled` to detect when an alternate commitment has taken over, and then go to the regular local close flow (using that commitment). This also clears up a potential confusion where we use `WatchFundingSpent` for two very different purposes: spending the output of the funding tx (original case), and double spending the funding tx (new use case).

Reply:
Done in https://github.com/ACINQ/eclair/pull/2274/commits/d5d0e36ccdf018dff3d845aa9cec8beb84089b97, I think it's indeed better to use `WatchFundingConfirmed` instead of `WatchFundingSpent`, the "trick" is to update the underlying state (`DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED`) to "erase" obsolete previous txs when one confirms.

So now in each transition, we have:

- `WAIT_FOR_DUAL_FUNDING_CONFIRMED` -> `CLOSING`: we have put `WatchFundingConfirmed` on each transaction (and no `WatchFundingSpent`), so the handler for that event in `CLOSING` should correctly settle every case
- `WAIT_FOR_FUNDING_CREATED`->`OFFLINE`[->`SYNCING`]->`CLOSING`: we have put `WatchFundingConfirmed` on each transaction (and no `WatchFundingSpent`). Once one confirms, we update the inner `DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED` to have this commitments as the main one and put a `WatchFundingSpent`. When the remote spends the funding tx, we go through the normal flow of `handleRemoteSpentCurrent`. If an error occurs, we will spend our local commit from this confirmed funding tx.
- `WAIT_FOR_DUAL_FUNDING_CONFIRMED`->FSM stopped->`WAIT_FOR_INIT_INTERNAL`[->`OFFLINE`][->`SYNCING`]->`CLOSING`: we put back a `WatchFundingConfirmed` for each transaction, so we should be in the same state as the previous one

Reply:
We now also have the case `WAIT_FOR_FUNDING_CONFIRMED` -> `OFFLINE` -> `WatchFundingConfirmed` -> `WAIT_FOR_FUNDING_CONFIRMED`.

In that case when receiving `channel_reestablish` we put a `WatchFundingConfirmed` again, which will trigger immediately the corresponding handler in `WAIT_FOR_FUNDING_CONFIRMED` and move us to the next state.